<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Multi-Objective Optimization case study • fbar</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">fbar</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/Multi-Objective_Optimization_case_study.html">Multi-Objective Optimization case study</a>
    </li>
    <li>
      <a href="../articles/Tutorial.html">Tutorial</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/maxconway/fbar">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Multi-Objective Optimization case study</h1>
                        <h4 class="author">Max Conway</h4>
            
            <h4 class="date">2017-10-17</h4>
          </div>

    
    
<div class="contents">
<div id="multi-objective-optimization-tutorial" class="section level1">
<h1 class="hasAnchor">
<a href="#multi-objective-optimization-tutorial" class="anchor"></a>Multi Objective Optimization tutorial</h1>
<p>This tutorial is an in depth example of the use of this package in the context of an evolutionary optimization approach. It also shows the use of <code>gene_eval</code> to calculate a reaction speed from gene expressions.</p>
<p>First, we need to load the appropriate libraries. You may also need to load your optimization library of choice.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)</code></pre></div>
<pre><code>## Warning: Installed Rcpp (0.12.13) different from Rcpp used to build dplyr (0.12.11).
## Please reinstall dplyr to avoid random crashes or undefined behavior.</code></pre>
<pre><code>## Loading tidyverse: ggplot2
## Loading tidyverse: tibble
## Loading tidyverse: tidyr
## Loading tidyverse: readr
## Loading tidyverse: purrr
## Loading tidyverse: dplyr</code></pre>
<pre><code>## Conflicts with tidy packages ----------------------------------------------</code></pre>
<pre><code>## filter(): dplyr, stats
## lag():    dplyr, stats</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(stringr)
<span class="kw">library</span>(fbar)</code></pre></div>
<p>This code block loads a model, then extracts the list of genes from the model. The model takes the form of a tabular list of reactions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">"ecoli_core"</span>)
model &lt;-<span class="st"> </span>ecoli_core

genes_in_model &lt;-<span class="st"> </span>model$geneAssociation %&gt;%
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/stringr/topics/str_split">str_split</a></span>(<span class="st">'[()|&amp; ]+'</span>) %&gt;%
<span class="st">  </span><span class="kw">flatten_chr</span>() %&gt;%
<span class="st">  </span><span class="kw">discard</span>(is.na) %&gt;%
<span class="st">  </span><span class="kw">discard</span>(~<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/stringr/topics/str_length">str_length</a></span>(.x)==<span class="dv">0</span>)</code></pre></div>
<p>The evaluation function is where the actual metabolic simulations are performed. This has four main stages:</p>
<ul>
<li>The gene-reaction associations (<code>geneAssociation</code>) are evaluated in the context of which genes are present in this iteration (<code>genome</code>).</li>
<li>We conduct a round of FBA, optimizing for maximum biomass.</li>
<li>Having found the maximum biomass production, we fix the biomass at this value (+/-1%).</li>
<li>With the biomass value fixed, we the optimize to maximize the synthetic objective.</li>
</ul>
<p>The technique of fixing the biomass followed by maximizing the synthetic objective is important because there could still be slack in the model after the first optimization stage, and we wish to have a reliable synthetic objective estimate.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">evaluation_function &lt;-<span class="st"> </span>function(genome){
  
  res &lt;-<span class="st"> </span>model %&gt;%
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">activation =</span> <span class="kw"><a href="../reference/gene_eval.html">gene_eval</a></span>(geneAssociation, <span class="kw">names</span>(genome), genome),
           <span class="dt">activation =</span> <span class="kw">coalesce</span>(activation, <span class="dv">1</span>),
           <span class="dt">uppbnd =</span> uppbnd*activation,
           <span class="dt">lowbnd =</span> lowbnd*activation) %&gt;%
<span class="st">    </span><span class="kw"><a href="../reference/find_fluxes_df.html">find_fluxes_df</a></span>(<span class="dt">do_minimization =</span> <span class="ot">FALSE</span>) %&gt;%
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">lowbnd =</span> <span class="kw">ifelse</span>(abbreviation==<span class="st">'Biomass_Ecoli_core_w/GAM'</span>, flux*<span class="fl">0.99</span>, lowbnd),
           <span class="dt">uppbnd =</span> <span class="kw">ifelse</span>(abbreviation==<span class="st">'Biomass_Ecoli_core_w/GAM'</span>, flux*<span class="fl">1.01</span>, uppbnd),
           <span class="dt">obj_coef =</span> <span class="dv">1</span>*(abbreviation==<span class="st">'EX_ac(e)'</span>)) %&gt;%
<span class="st">    </span><span class="kw"><a href="../reference/find_fluxes_df.html">find_fluxes_df</a></span>(<span class="dt">do_minimization =</span> <span class="ot">FALSE</span>)
  
  <span class="kw">return</span>(<span class="kw">list</span>(<span class="dt">bm =</span> <span class="kw">filter</span>(res, abbreviation==<span class="st">'Biomass_Ecoli_core_w/GAM'</span>)$flux, 
              <span class="dt">synth =</span> <span class="kw">filter</span>(res, abbreviation==<span class="st">'EX_ac(e)'</span>)$flux))
}</code></pre></div>
<p>Non-domination sorting is the first stage of the selection procedure in NSGA-II. The code might be quite opaque, but the idea is as follows:</p>
<ul>
<li>We perform an inner_join in order to compare every point against every other point.</li>
<li>For each point (<code>id.x</code>), we see if there exists any second point (<code>id.y</code>) that has a higher value than it in all objectives. Where such a second point exists, we term the original point ‘dominated’.</li>
<li>We find the set of points which have no dominating point, and term this the first non-dominated front.</li>
<li>We repeat this procedure, but ignoring points in the first non-dominated front, to find the second on-dominated front, and so on.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">non_dom_sort &lt;-<span class="st"> </span>function(input){
  input_long &lt;-<span class="st"> </span>input %&gt;%
<span class="st">    </span><span class="kw">gather</span>(property, value, -id) %&gt;%
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">front=</span><span class="ot">NA</span>)
  
  currentfront &lt;-<span class="st"> </span><span class="dv">1</span>
  
  while(<span class="kw">any</span>(<span class="kw">is.na</span>(input_long$front))){
    
    input_long &lt;-<span class="st"> </span>input_long %&gt;%
<span class="st">      </span><span class="kw">inner_join</span>(.,., <span class="dt">by=</span><span class="st">'property'</span>) %&gt;%
<span class="st">      </span><span class="kw">group_by</span>(id.x,id.y) %&gt;%
<span class="st">      </span><span class="kw">mutate</span>(<span class="dt">dominance =</span> <span class="kw">ifelse</span>(<span class="kw">all</span>(value.x&gt;=value.y), 
                                <span class="st">'xdomy'</span>, 
                                <span class="kw">ifelse</span>(<span class="kw">all</span>(value.y&gt;=value.x), 
                                       <span class="st">'ydomx'</span>, 
                                       <span class="st">'nondom'</span>
                                       )
                                )
      ) %&gt;%
<span class="st">      </span><span class="kw">group_by</span>(id.x) %&gt;%
<span class="st">      </span><span class="kw">mutate</span>(<span class="dt">front =</span> <span class="kw">ifelse</span>(<span class="kw">all</span>(dominance[<span class="kw">is.na</span>(front.y)] %in%<span class="st"> </span><span class="kw">c</span>(<span class="st">'xdomy'</span>, <span class="st">'nondom'</span>)), 
                            <span class="kw">pmin</span>(currentfront, front.x, <span class="dt">na.rm=</span><span class="ot">TRUE</span>), 
                            <span class="ot">NA</span>
                            )
      ) %&gt;%
<span class="st">      </span><span class="kw">group_by</span>(<span class="dt">id =</span> id.x, <span class="dt">property =</span> property, front, <span class="dt">value =</span> value.x) %&gt;%
<span class="st">      </span><span class="kw">summarise</span>()
    
    currentfront &lt;-<span class="st"> </span>currentfront +<span class="st"> </span><span class="dv">1</span>
  }
  
  <span class="kw">return</span>(
    input_long %&gt;%
<span class="st">      </span><span class="kw">spread</span>(property, value)
  )
  
}</code></pre></div>
<p>The second part of the NSGA-II evaluation procedure is finding the crowding distance. This is used to break ties between points in the same non-dominated front. In for each front, for each dimension, this function sorts the points into order along the dimension, and finds the normalized distance between the proceeding point and succeeding point. These values are summed up across each dimension to find the value for the point.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">crowding_distance &lt;-<span class="st"> </span>function(input){
  <span class="kw">return</span>(
    input %&gt;%
<span class="st">      </span><span class="kw">gather</span>(property, value, -id, -front) %&gt;%
<span class="st">      </span><span class="kw">group_by</span>(front, property) %&gt;%
<span class="st">      </span><span class="kw">arrange</span>(value) %&gt;%
<span class="st">      </span><span class="kw">mutate</span>(<span class="dt">crowding =</span> (<span class="kw">lead</span>(value)-<span class="kw">lag</span>(value))/(<span class="kw">max</span>(value)-<span class="kw">min</span>(value)),
             <span class="dt">crowding =</span> <span class="kw">ifelse</span>(<span class="kw">is.na</span>(crowding),<span class="ot">Inf</span>, crowding)) %&gt;%
<span class="st">      </span><span class="kw">group_by</span>(id) %&gt;%
<span class="st">      </span><span class="kw">mutate</span>(<span class="dt">crowding =</span> <span class="kw">sum</span>(crowding)) %&gt;%
<span class="st">      </span><span class="kw">spread</span>(property, value)
  )
}</code></pre></div>
<p>This is the genetic loop of the algorithm. It is explained by code comments, but follows a normal pattern of evaluating, sorting, selecting from and mutating the population.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">start_genome &lt;-<span class="st"> </span><span class="kw">set_names</span>(<span class="kw">rep_along</span>(genes_in_model, <span class="ot">TRUE</span>), genes_in_model)
pop &lt;-<span class="st"> </span><span class="kw">list</span>(start_genome)

popsize =<span class="st"> </span><span class="dv">50</span>

for(i in <span class="dv">1</span>:<span class="dv">50</span>){
  results &lt;-<span class="st"> </span><span class="kw">map_df</span>(pop, evaluation_function) %&gt;%<span class="st"> </span><span class="co"># Evaluate all the genomes</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">bm=</span><span class="kw">signif</span>(bm), <span class="dt">synth=</span><span class="kw">signif</span>(synth)) %&gt;%<span class="st"> </span><span class="co"># Round results</span>
<span class="st">    </span><span class="kw">unique</span>() %&gt;%<span class="st"> </span><span class="co"># Throw away duplicates</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">id =</span> <span class="dv">1</span>:<span class="kw">n</span>()) %&gt;%<span class="st"> </span><span class="co"># label the results</span>
<span class="st">    </span><span class="kw">sample_frac</span>() %&gt;%<span class="st"> </span><span class="co"># Shuffle</span>
<span class="st">    </span><span class="kw">non_dom_sort</span>() %&gt;%<span class="st"> </span><span class="co"># Find the non-dominated fronts</span>
<span class="st">    </span>crowding_distance %&gt;%<span class="st"> </span><span class="co"># Find the crowding distances</span>
<span class="st">    </span><span class="kw">arrange</span>(front, <span class="kw">desc</span>(crowding)) <span class="co"># Sort by front, breaking ties by crowding distance</span>
  
  selected &lt;-<span class="st"> </span>results %&gt;%
<span class="st">    </span><span class="kw">filter</span>(<span class="kw">row_number</span>() &lt;=<span class="st"> </span>popsize/<span class="dv">2</span>) %&gt;%<span class="st"> </span><span class="co"># Keep the best half of the population</span>
<span class="st">    </span><span class="kw">getElement</span>(<span class="st">'id'</span>)
  
  kept_pop &lt;-<span class="st"> </span>pop[selected]
  altered_pop &lt;-<span class="st"> </span>kept_pop %&gt;%
<span class="st">    </span><span class="kw">sample</span>(popsize-<span class="kw">length</span>(selected), <span class="ot">TRUE</span>) %&gt;%<span class="st"> </span><span class="co"># Select a random portion of the population as parents</span>
<span class="st">    </span><span class="kw">map</span>(function(genome){<span class="kw">xor</span>(genome, <span class="kw">runif</span>(<span class="kw">length</span>(genome))&gt;<span class="fl">0.98</span>)}) <span class="co"># Mutate a small number of genes from the parent population.</span>
  
  pop &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">c</span>(kept_pop, altered_pop)) <span class="co"># Combine the ofspring and parent populations</span>
}</code></pre></div>
<p>Now that we have results, the set of all non-dominated points, known as the Pareto Front. This describes the tradeoff between biomass production and our synthetic objective.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)

results %&gt;%
<span class="st">  </span><span class="kw">filter</span>(front==<span class="dv">1</span>) %&gt;%
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">x=</span>bm, <span class="dt">y=</span>synth, <span class="dt">colour=</span>front)) +<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_point">geom_point</a></span>() +<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_path">geom_step</a></span>(<span class="dt">direction=</span><span class="st">'vh'</span>) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/scale_continuous">scale_x_log10</a></span>() +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/scale_continuous">scale_y_log10</a></span>() +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggtheme">theme_bw</a></span>()</code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#multi-objective-optimization-tutorial">Multi Objective Optimization tutorial</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Max Conway.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
